fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,176 @@
+                                .addLogInfo(LogMessageKeys.SUBSPACE_KEY, ByteArrayUtil2.loggable(lockSubspace.getKey()))
+                                .addLogInfo(LogMessageKeys.UUID, sessionId);
+                    }
+                });
+            }
+        });
+    }
+
+    private CompletionStage<Void> takeSessionLock(Transaction tr) {
+        setLockSessionId(tr);
+        setLockSessionTime(tr);
+        return AsyncUtil.DONE;
+    }
+
+    // Check and renew the lock when the session in being used.
+    // TODO: Maybe the time should be updated even if the work is failed. For example, online indexer may fail
+    // in the first a few transactions to find the optimal number of records to scan in one transcation.
+    <T> Function<FDBRecordContext, T> workInSession(
+            @Nonnull Function<? super FDBRecordContext, ? extends T> work) {
+        return context -> {
+            context.asyncToSync(FDBStoreTimer.Waits.WAIT_CHECK_SYNC_SESSION, checkLock(context));
+            T result = work.apply(context);
+            setLockSessionTime(context.ensureActive());
+            return result;
+        };
+    }
+
+    <T> Function<? super FDBRecordContext, CompletableFuture<? extends T>> workInSessionAsync(
+            @Nonnull Function<? super FDBRecordContext, CompletableFuture<? extends T>> work) {
+        return context -> checkLock(context)
+                .thenCompose(vignore -> work.apply(context))
+                .thenApply(result -> {
+                    setLockSessionTime(context.ensureActive());
+                    return result;
+                });
+    }
+
+    @Nonnull
+    UUID getSessionId() {
+        return sessionId;
+    }
+
+    private CompletableFuture<Void> checkLock(FDBRecordContext context) {
+        return getLockSessionId(context.ensureActive())
+                .thenCompose(lockSessionId -> {
+                    if (!sessionId.equals(lockSessionId)) { // sessionId is nonnull while lockSessionId is nullable
+                        throw new SynchronizedSessionExpiredException("Failed to continue the session")
+                                .addLogInfo(LogMessageKeys.SUBSPACE_KEY, ByteArrayUtil2.loggable(lockSubspace.getKey()))
+                                .addLogInfo(LogMessageKeys.UUID, sessionId)
+                                .addLogInfo("lockSessionId", lockSessionId);
+                    }
+                    return AsyncUtil.DONE;
+                });
+    }
+
+    void close(FDBRecordContext context) {
+        context.ensureActive().clear(lockSubspace.pack(), ByteArrayUtil.strinc(lockSubspace.pack()));
+    }
+
+    private CompletableFuture<UUID> getLockSessionId(@Nonnull Transaction tr) {
+        return tr.get(lockSessionIdSubspaceKey)
+                .thenApply(value -> value == null ? null : Tuple.fromBytes(value).getUUID(0));
+    }
+
+    private void setLockSessionId(@Nonnull Transaction tr) {
+        tr.set(lockSessionIdSubspaceKey, Tuple.from(sessionId).pack());
```

Comment:
Maybe this tuple should include the least time.  I think this was discussed at some point as a safety net for a client trying to join the synchronized session but with a different lease time.  In fact, maybe it would make sense that joining a session does not even allow you to specify the lease time.